endorsement/dao/shared_drive.py

```diff
@@ -70,48 +104,311 @@ def get_drive_quota(a):
+        # only add provisions  and quanties to itbill models
+        # for the given product sys_id
+        # (could it be the case that the product sys_id is not unique?)
+
+    def handle_new_drives(self, drive_ids, id_GoogleDriveState, default_quota):
+        """
+        Create shared drive record with implicit provisioned date.
+
+        This is expected on the first day of acceptance lifecycle as we get
+        everything previously existing bootstrapped. After that, in theory, all
+        the new shared drives will be created via PRT and thus this shouldn't
+        happen.
+        """
+        for drive_id in drive_ids:
+            drive_states = id_GoogleDriveState[drive_id]
+
+            # TODO: probably wrap remainder of loop in a transaction...
+
+            # create SharedDriveRecord + SharedDrive + Members/Roles
+            sdr: SharedDriveRecord = load_shared_drives(drive_states)
+            if sdr is None:  # diagnostic
+                print(
+                    f"Got None back when providing {len(drive_states)} "
+                    "records to load_shared_Drive"
+                )
+                print(
+                    "{drive_id=} Member roles: "
+                    f"{set(X.role for X in drive_states)}"
+                )
+
+            if sdr is None:
+                # examples:
+                # drive_id 0APhs8SozJRR3Uk9PVA; all members in SUSPENDED role
+                #   if no drive members have an "organizer" role the
+                #   SharedDriveRecord is not created
+
+                # In these cases PRT has no administrator to alert
+                # TODO: confirm with dsnorton this should be silently ignored
+                continue
+
+            quota = sdr.shared_drive.drive_quota
+            if quota.quota_limit == default_quota:
+                continue
+
+            sub = get_subscription(sdr)
+            if sub is None:
+                set_drive_quota(drive_id=drive_id, quota=default_quota)
+                quota.quota_limit = default_quota
+                quota.save()
+
+    def handle_missing_drives(self, missing_drive_ids, default_quota):
+        """
+        For drives that no longer exist as reported by MSCA.
+
+        This means the drive no longer exists or no longer has managers.
+
+        If deleted it is expected the related SharedDriveRecord would have
+        is_deleted set to True.
+
+        The last manager departing would happen out of band.
+        """
+        # TODO: should we be logging this or performing any other actions?
+        records = SharedDrive.objects.filter(drive_id__in=missing_drive_ids)
+        for record in records:
+            record.is_deleted = True
+
+        SharedDrive.objects.bulk_update(records, fields=["is_deleted"])
+
+    def reconcile_existing_drives(
+        self, drives, id_GoogleDriveState, subsidized_quota
+    ):
+        # helper functions that probably need to be moved to a DAO file
```

Comment:
Ignore this function - I quickly put some stubs together but it's not even remotely complete